## 
1. This github repository is for the [**A custom library construction method for super-resolution ribosome profiling in Arabidopsis**](https://www.biorxiv.org/content/10.1101/2022.07.14.499987v1.abstract) paper currently on biorxiv and is accepted for publication in Plant Methods. Although this paper is about how to create Ribo-seq library, we provide a simple pipeline for the quality control here. Please cite this paper if you find the information provided here is useful.  
2. You can run the code on locally (PC/Mac), linux server or linux cluster with properly installed/loaded software packages.  
3. You have to decide the $INPUT and $OUTPUT paths for each step.  
4. Here we provide a small example from the sample NEB1. The FASTA and transcriptome GTF files can be obtained from [**TAIR**](https://www.arabidopsis.org). Use Chrome browser. Connect as a guest.  
5. This is beyond the scope of the paper, but more info about downstream analysis of Riboseq can be found [**HERE**](https://github.com/hsinyenwu/Ribo-seq-informatics-2022). For visualizing the Ribo-seq reads on individual transcripts, please try out [**RiboPlotR**](https://github.com/hsinyenwu/RiboPlotR).
6. If run with > R (v. 4.2), there is a known issue for RiboseQC (from the GenomicFeatures package). A modified version of RiboseQC is available [**HERE**](https://github.com/hsinyenwu/RiboseQC_R4.2.1/blob/master/R/riboseqc.R). Use ```install_github(repo = "hsinyenwu/RiboseQC_R4.2.1")``` then load the package with ```library(RiboseQC)```.
