### Protocol and bioinformatic analysis pipeline for high-coverage and super-resolution ribosome profiling  

#### Information included
1. New protocol  
2. Catalog numbers for reagents
3. How to further improve the method (e.g., only cut 25-30 region-> may increase mapped read numbers beacuse 24-25 lots of rRNA contamination, add UMI)
4. Smaller peak ~21 nt also mapped to CDS
5. 29 nt reads show strong stop codon
6. Best still 28 nt  
7. Code for analysis and visualization

#### Where is the code?
Linux: 
Mac:

#### What to do?
1. Bam files select reads from 25 to 30 nts
2. Get same read numbers
3. Sort and index
4. RiboTaper analysis
5. Plot several genes
6. RiboseQC output
7. ORFquant analysis?
8. Correlation between samples

### Figures:
1. Workflow for experiment and analysis
2. Meta plot (28 nt and 21 nt), read length distribution, read feature distribution, read P-site coverage, per frame coverage
3. Table: number of genes identified in NEB vs CTRL vs Root vs Shoot data
4. uORF and sORF overlap between NEB and CTRL, Gene examples















